File: src/test/java/com/github/pagviewer/nativebridge/FakePagNativeLibrary.java
package com.github.pagviewer.nativebridge;

import java.nio.ByteBuffer;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicInteger;

final class FakePagNativeLibrary implements PagNativeLibrary {
    private final AtomicLong handles = new AtomicLong(1);
    private final int width;
    private final int height;
    private final int frames;
    private final float frameRate;
    private final byte[] pixels;
    private final boolean[] changedFrames;
    private final AtomicInteger readFrameCalls = new AtomicInteger();

    FakePagNativeLibrary(int width, int height, int frames, float frameRate, byte[] pixels) {
        this(width, height, frames, frameRate, pixels, null);
    }

    FakePagNativeLibrary(int width, int height, int frames, float frameRate, byte[] pixels, boolean[] changedFrames) {
        this.width = width;
        this.height = height;
        this.frames = frames;
        this.frameRate = frameRate;
        this.pixels = pixels;
        this.changedFrames = changedFrames;
    }

    @Override
    public long loadFile(byte[] bytes, String filePath) {
        return bytes.length == 0 ? 0 : handles.getAndIncrement();
    }

    @Override
    public long createDecoder(long composition, float maxFrameRate, float scale) {
        return composition == 0 ? 0 : handles.getAndIncrement();
    }

    @Override
    public int compositionWidth(long composition) {
        return width;
    }

    @Override
    public int compositionHeight(long composition) {
        return height;
    }

    @Override
    public int width(long decoder) {
        return width;
    }

    @Override
    public int height(long decoder) {
        return height;
    }

    @Override
    public int frameCount(long decoder) {
        return frames;
    }

    @Override
    public float frameRate(long decoder) {
        return frameRate;
    }

    @Override
    public boolean checkFrameChanged(long decoder, int frameIndex) {
        return changedFrames == null || changedFrames[Math.floorMod(frameIndex, changedFrames.length)];
    }

    @Override
    public boolean readFrame(long decoder, int frameIndex, ByteBuffer destination, int rowBytes) {
        readFrameCalls.incrementAndGet();
        destination.clear();
        destination.put(pixels);
        destination.flip();
        return true;
    }

    int readFrameCalls() {
        return readFrameCalls.get();
    }

    @Override
    public void release(long handle) {
    }
}

File: src/test/kotlin/com/github/pagviewer/nativebridge/FakePagNativeLibrary.kt
package com.github.pagviewer.nativebridge

import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

internal class FakePagNativeLibrary(
    private val width: Int,
    private val height: Int,
    private val frames: Int,
    private val frameRate: Float,
    private val pixels: ByteArray,
    private val changedFrames: BooleanArray? = null
) : PagNativeLibrary {
    private val handles = AtomicLong(1)
    private val readFrameCalls = AtomicInteger()

    override fun loadFile(bytes: ByteArray, filePath: String): Long =
        if (bytes.isEmpty()) 0 else handles.getAndIncrement()

    override fun createDecoder(composition: Long, maxFrameRate: Float, scale: Float): Long =
        if (composition == 0L) 0 else handles.getAndIncrement()

    override fun compositionWidth(composition: Long): Int = width

    override fun compositionHeight(composition: Long): Int = height

    override fun width(decoder: Long): Int = width

    override fun height(decoder: Long): Int = height

    override fun frameCount(decoder: Long): Int = frames

    override fun frameRate(decoder: Long): Float = frameRate

    override fun checkFrameChanged(decoder: Long, frameIndex: Int): Boolean {
        val changed = changedFrames ?: return true
        return changed[Math.floorMod(frameIndex, changed.size)]
    }

    override fun readFrame(decoder: Long, frameIndex: Int, destination: ByteBuffer, rowBytes: Int): Boolean {
        readFrameCalls.incrementAndGet()
        destination.clear()
        destination.put(pixels)
        destination.flip()
        return true
    }

    fun readFrameCalls(): Int = readFrameCalls.get()

    override fun release(handle: Long) {}
}
